Headers

shahifaqeer opened the pull request.

shahifaqeer committed:
Add getHeaders function to return 4 byte packed headers as a string

shahifaqeer committed:
fixed format string and last byte for flags and nonce

assume input flags and nonce are both integers to be packed
together into a byte after bit shifting

shahifaqeer committed:
generate nonce and flags

need to get sequence numbers and session IDs

shahifaqeer committed:
add an UnAssembler class

receives a full frame and unassembles it to data
and headers. Then, unpacks the headers to set appropriate
values

gsathya commented on htpt/frame.py:
```diff
@@ -5,6 +5,7 @@
 
 import threading
 import numpy as np
+import struct as struct
```
`import struct` is enough

gsathya commented on htpt/frame.py:
```diff
@@ -142,9 +143,50 @@ class Assembler:
+  
+  def generateNonce(self)
+    """Generate a random integer value between 0 and 16"""
+    nonce = np.random.randint(0,15)
```
no need for numpy. python can do this

shahifaqeer replied:
using random.randint() instead

gsathya commented on htpt/frame.py:
```diff
@@ -142,9 +143,50 @@ class Assembler:
+    sessionID = None
+    return sessionID
+
+  def getHeaders(self, seqNum, sessionID, flags, nonce):
```
getHeaders should call getSeqNum, getSessionID, generateFlags. The values shouldn't be passed as arguments

gsathya commented on htpt/frame.py:
```diff
@@ -142,9 +143,50 @@ class Assembler:
+
+    """
+    # 16-bit sequence num | 8-bit session ID | 4-bit flag | 4-bit nonce
+    # unsigned short (H) | unsigned char (B) | unsigned char (B) packed
```
These comments should go in the docstring

gsathya commented on htpt/frame.py:
```diff
@@ -142,9 +143,50 @@ class Assembler:
+      flags = '0000'
+    return flags
+  
+  def generateNonce(self)
```
missing ":" at the end

gsathya commented on htpt/frame.py:
```diff
@@ -142,9 +143,50 @@ class Assembler:
+    nonce = np.random.randint(0,15)
+    return nonce
+
+  def getSeqNum(self)
```
missing ":" at the end

gsathya commented on htpt/frame.py:
```diff
@@ -142,9 +143,50 @@ class Assembler:
+    seqNum = None
+    return seqNum
+
+  def getSessionID(self)
```
missing ":" at the end

shahifaqeer committed:
fixed sathya's comments

fixed typos missing :
shifted getHeaders() comments to docstring
calling generate/get functions in getHeaders()

TODO: seqNum, sessionID, more_data in flags

shahifaqeer committed:
fix seqNum generator; change to Encoder/Decoder

gsathya commented:
@ben-jones can you please review this?

ben-jones commented on htpt/frame.py:
```diff
@@ -12,33 +14,22 @@
-  _initialized = False
+  self.seqNum = 0
+  self.initialized = False
+  self._lock = threading.Lock()
```
Why did we make these instance variables instead of class variables? It works either way, but now you need to create an instance of the SeqNumber class before you can do anything with sequence numbers.

gsathya replied:
I think that's fine.

ben-jones commented on htpt/frame.py:
```diff
@@ -12,33 +14,22 @@
+    return self.seqNum
 
 #make this methods available without having to access the class
 getSequenceAndIncrement = SeqNumber.getSequenceAndIncrement
```
If we are switching everything from class variables (no instantiation) to instance variables (an object is needed) then this line is no longer valid. If we still want to hide the implementation details, we could create a wrapper function which does lazy instantiation (create an object the first time you need it) of an object of the SeqNumber class. However, this line of code is invalid after the code changes.

gsathya replied:
Agreed. This should be removed. @shahifaqeer I removed this in my patch, not sure if you noticed it

shahifaqeer replied:
fixed.

gsathya replied:
where?

ben-jones commented on htpt/frame.py:
```diff
@@ -138,19 +129,112 @@ def flushBuffer(self):
+    
+    data = frame[4:]
+    for datum in data:
+      self.output = urlEncode.decode(datum)
```
What happened to the code to buffer the output as it came in? My code needed improvements, but it at least allowed us to queue packets. Also, this code overwrites the value of the buffer every time that you receive data.

gsathya replied:
I think buffers are going to do the buffering. The encoder/decoder should just encode or decode, not buffer. See htpt/buffers.py

ben-jones replied:
That totally makes sense. I have not looked at that code yet. Also, what is the next step for merging the changes?

gsathya replied:
Wait till Sarthak reviews the comments and revises the code. Then do the pull request dance again

gsathya commented:
I'm going to wait on merging till all the framing work is done. (All the TODOs are done) 

Please let me know if this is blocking anyone, I'll merge it off then. (Once the code is revised, ofc)

shahifaqeer committed:
removed global variable

getSequenceAndIncrement global variable was removed

gsathya commented:
@shahifaqeer, any updates on the TODOs?

shahifaqeer committed:
create a SessionID class to create and assign sessionIDs

shahifaqeer committed:
set more_data flag when data > BUFFER_SIZE

shahifaqeer committed:
encode functions reads **kwargs for more_data flag

TODO: decode(): if more_data is set in headers of received
packet, then send a pull request to the server to send the
rest of the buffer.

TODO: in buffers, only set more_data if data > BUFFER_SIZE.
if (data == BUFFER_SIZE) or (pull request and data > 0), then
set more_data = 0 and flushBuffer().

shahifaqeer committed:
ignore .ropeproject

gsathya commented:
@shahifaqeer, Is this done?

shahifaqeer committed:
fixed typos; shifted initialized variables to init function in class SeqNumber

shahifaqeer committed:
fixed some spacing [pep8]

shahifaqeer committed:
fix header struct error

header struct was not packing correctly as session ID was
set to None by default. Instead set default value to 0 so
that integer can be converted.

shahifaqeer committed:
fix flags refernced before assignment error in generateFlags

shahifaqeer committed:
decode incoming packet to frame before splitting into headers + data

we assemble a frame = headers+data, and then encode it to a packet
using urlEncode.encode() before sending it to flush().
Similarly, we should decode incoming packet to a frame using
urlEncode.decode before splitting into headers+data.

shahifaqeer committed:
rename classes to Assemble and Disassemble to avoid confusion with urlEncode

shahifaqeer committed:
fix errors and check Assemble and Disassemble with sequence numbers and data

shahifaqeer committed:
remove Framer.recvFrame and Framer.flushBuffer functions

these functions should be implemented in buffer.py. Framer is only
responsible for assembly/disassembly while maintaining correct
sequence numbers.

shahifaqeer commented:
Done. Ready for merge.
Tested Assemble and Disassemble - they work perfectly.

gsathya commented:
Can you please rebase this on my current master? Thanks!

shahifaqeer committed:
merged: fixed import statements

gsathya commented:
and the unit tests?

shahifaqeer committed:
Merge branch 'master' of github.com:gsathya/htpt into headers

shahifaqeer committed:
irearchitect buffers, implement recvData with seqNum check

shift recvData() functionality from Framer to Buffer.
define buffer as a list. new data is inserted at an index in this list
and flushed to upper layers in order.

TODO: fix addData()

shahifaqeer committed:
start adding buffers to framing

shahifaqeer committed:
make note to remove addData function

shahifaqeer committed:
Merge branch 'buffers' into headers

shahifaqeer committed:
fixed callback to flush() after recvData

shahifaqeer committed:
removed encode/decode from assembly/disassembly

shahifaqeer committed:
get/set SessionID(); option for SYN flag

shahifaqeer committed:
TODO: interpret flags at receiver and take action

shahifaqeer committed:
implement class SessionID

sessionID is incremented at returned when a new client
joins. This sessionID should then be passed to
Assemble.setSessionID or Disassemble.setSessionID

shahifaqeer committed:
new frame and buffer tests

shahifaqeer committed:
fixed bugs

self.seqNum was getting reassigned recursively
flags were being retrieved as integer, changed it to binary string

shahifaqeer committed:
unit tests for assembly and disassembly

gsathya commented:
@ben-jones Yo can we haz code review?

ben-jones commented on htpt/buffers.py:
```diff
@@ -17,14 +40,80 @@ def addData(self, data):
+        dataToSend = dataToSend + len(self.buffer[index])
+    #if we have enough data to send up, pass it to the callback
+    #function
+    if dataToSend > MIN_SIZE_TO_PASS_UP:
```
Since we are dropping the min size to pass up, a lot of this code needs to be removed. remove lines 98 to 109 since they shouldn't do anything and the while loop on line 112 takes care of everything now.

ben-jones commented on htpt/buffers.py:
```diff
@@ -1,15 +1,38 @@
   def addCallback(self, callback):
     self.callback = callback
 
   def addData(self, data):
```
I think the original point of this function was to avoid having to write all the mod MAX_SEQ_NUM code every time it was needed. This function no longer appears necessary and should be deleted.

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+    flags = "".join(flag_list)
+    return flags
+
+  def generateNonce(self):
```
I realize that I suggested we use a nonce, but it turns out that I am an idiot. Nonces need to have a low probability of reoccurring, which the small range for this nonce does not provide. To be less confusing, it would be better to delete this function.

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+    self.flags = self.generateFlags(**kwargs)
+    self.nonce = self.generateNonce()
+
+    flags_and_nonce = (int(self.flags,2) << 4) | self.nonce
```
Sorry, the code above makes more sense now

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+
+    headers = self.getHeaders(**kwargs)
+    frame = headers+data
+    self.output = frame
```
why don't you just return the frame as a string? Otherwise, you could end up accidentally overwriting data and it creates a simpler API

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+    self.retrieveHeaders(headers)
+
+    data = frame[4:]
+    self.output = data
```
I don't really understand the differences between the buffers and self.output. Can you stick to just sending the data to the buffers or returning it to the user.

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+    # At main: if flags = '1000' i.e. more_data, then send pull_request to
+    # server for more data
+    # At main: if SYN flag is set then set/get session ID
+    self.nonce = headerTuple[2] & mask
```
Again, my apologies, but we won't need the nonce section at all.

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
   def flush(self):
-    # TODO: send self.output to the interwebz
+    # send self.output to Tor
+    print self.output
```
Can you just return the data to the calling function.? Also, what is self.output and why isn't there any reference to a buffer?

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+
+
+class Disassemble:
+  """Class to Disassemble a decoded packet into headers+data before sending to buffers"""
```
Will you please add a setter method to set a callback for this class or take it as a parameter in the constructor? Callbacks are pretty simple, so it could just be something like
def setCallback(self, callback):
     self.callback = callback
You would use this callback to send data up to the upper layers. So when the data gets flushed from the buffer, it just calls the callback, like self.callback(inOrderData). If you include this in the constructor, then you won't need a flush method, and you can just set the callback for the buffer to the callback you get in the constructor.

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+    # TODO in main(): encode self.output to a packet and send output packet to flush()
+
+
+class Disassemble:
```
It would be extremely nice to have an initialization function here. Otherwise, I am now sure how I am going to be able to get consistent , non repeated session ids

ben-jones commented on htpt/frame.py:
```diff
@@ -4,152 +4,204 @@
+    self.initialized = True
+    self._lock = threading.Lock()
+
+  def getSessionIDAndIncrement(self):
```
How do you foresee this class being used? Right now, everything uses instance variables, which means 1) you need an object to be able to do anything with this class and 2) you will have different session IDs across different objects. This needs to either use static methods (see my seq num code) or it needs to not be an object (you would then need code to make sure that everything was properly initialized)

shahifaqeer committed:
removed extra functions; added callback option

shahifaqeer committed:
removed nonce; fixed flags; interfaced with buffers

shahifaqeer committed:
fixed testsuite with changes to frame.py

no Nonce functionality
dummycallback option for receive buffers
flags are now integers instead of strings
